src/main/java/org/apache/datasketches/quantilescommon/KolmogorovSmirnov.java

```diff
@@ -0,0 +1,167 @@
+    final long[] pCumWtsArr = p.getCumulativeWeights();
+    final long[] qCumWtsArr = q.getCumulativeWeights();
+    final int pSamplesArrLen = pSamplesArr.length;
+    final int qSamplesArrLen = qSamplesArr.length;
+
+    final double n1 = sketch1.getN();
+    final double n2 = sketch2.getN();
+
+    double deltaHeight = 0;
+    int i = 0;
+    int j = 0;
+
+    while ((i < pSamplesArrLen - 1) && (j < qSamplesArrLen - 1)) {
+      deltaHeight = Math.max(deltaHeight, Math.abs(pCumWtsArr[i] / n1 - qCumWtsArr[j] / n2));
+      if (pSamplesArr[i] < qSamplesArr[j]) {
+        i++;
+      } else if (qSamplesArr[j] < pSamplesArr[i]) {
+        j++;
+      } else {
+        i++;
+        j++;
+      }
+    }
+
+    deltaHeight = Math.max(deltaHeight, Math.abs(pCumWtsArr[i] / n1 - qCumWtsArr[j] / n2));
+    return deltaHeight;
+  }
+
+  /**
+   * Computes the adjusted delta height threshold for the <i>kolmogorovSmirnovTest(...)</i> method.
+   * This adjusts the computed threshold by the error epsilons of the two given sketches.
+   * The two sketches must be of the same primitive type, double or float.
+   * This will not work with the REQ sketch.
+   * @param sketch1 first Input QuantilesAPI sketch
+   * @param sketch2 second Input QuantilesAPI sketch
+   * @param tgtPvalue Target p-value. Typically .001 to .1, e.g., .05.
+   * @return the adjusted threshold to be compared with the raw delta area.
+   */
+  public static double computeKSThreshold(final QuantilesAPI sketch1,
+                                          final QuantilesAPI sketch2,
+                                          final double tgtPvalue) {
+    final double r1 = sketch1.getNumRetained();
+    final double r2 = sketch2.getNumRetained();
+    final double alpha = tgtPvalue;
+    final double alphaFactor = Math.sqrt(-0.5 * Math.log(0.5 * alpha));
+    final double deltaAreaThreshold = alphaFactor * Math.sqrt((r1 + r2) / (r1 * r2));
+    final double eps1 = sketch1.getNormalizedRankError(false);
+    final double eps2 = sketch2.getNormalizedRankError(false);
+    return deltaAreaThreshold + eps1 + eps2;
+  }
+
+  /**
+   * Performs the Kolmogorov-Smirnov Test between two QuantilesAPI sketches.
+   * Note: if the given sketches have insufficient data or if the sketch sizes are too small,
+   * this will return false. The two sketches must be of the same primitive type, double or float.
+   * This will not work with the REQ sketch.
+   * @param sketch1 first Input QuantilesAPI
+   * @param sketch2 second Input QuantilesAPI
+   * @param tgtPvalue Target p-value. Typically .001 to .1, e.g., .05.
+   * @return Boolean indicating whether we can reject the null hypothesis (that the sketches
+   * reflect the same underlying distribution) using the provided tgtPValue.
+   */
+  public static boolean kolmogorovSmirnovTest(final QuantilesAPI sketch1,
+      final QuantilesAPI sketch2, final double tgtPvalue) {
+
+    final double delta = isDoubleType(sketch1, sketch2)
```

Comment:
Kind of unfortunate we can't mix types :(

Reply:
In C++, yes, in Java, no.